core/src/test/java/com/datastax/oss/driver/internal/core/context/DefaultDriverContextTest.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.context;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.internal.core.protocol.Lz4Compressor;
+import com.datastax.oss.driver.internal.core.protocol.SnappyCompressor;
+import com.datastax.oss.protocol.internal.Compressor;
+import com.datastax.oss.protocol.internal.NoopCompressor;
+import com.tngtech.java.junit.dataprovider.DataProvider;
+import com.tngtech.java.junit.dataprovider.DataProviderRunner;
+import io.netty.buffer.ByteBuf;
+import java.util.Optional;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(DataProviderRunner.class)
+public class DefaultDriverContextTest {
+
+  private DefaultDriverContext buildMockedContext(Optional<String> compressionOption) {
+
+    DriverExecutionProfile defaultProfile = mock(DriverExecutionProfile.class);
+    when(defaultProfile.isDefined(DefaultDriverOption.PROTOCOL_COMPRESSION))
+        .thenReturn(compressionOption.isPresent());
+    compressionOption.ifPresent(
+        compression -> {
+          when(defaultProfile.getString(DefaultDriverOption.PROTOCOL_COMPRESSION))
+              .thenReturn(compression);
+        });
+    return MockedDriverContextFactory.defaultDriverContext(Optional.of(defaultProfile));
+  }
+
+  private void doCreateCompressorTest(Optional<String> configVal, Class<?> expectedClz) {
+
+    DefaultDriverContext ctx = buildMockedContext(configVal);
+    Compressor<ByteBuf> compressor = ctx.getCompressor();
+    assertThat(compressor).isNotNull();
+    assertThat(compressor).isInstanceOf(expectedClz);
+  }
+
+  @Test
+  @DataProvider({"lz4", "lZ4", "Lz4", "LZ4"})
```

Comment:
LGTM, but to make it even more maintanable maybe you could coalesce it to one test?
```
@DataProvider({"lz4", "lZ4", "Lz4", "LZ4","snappy", "SNAPPY", "sNaPpY", "SNapPy","none", "NONE", "NoNe", "nONe" })
public void should_create_compressor()
```

Reply:
Unfortunately you need to match up an input name with an expected compressor class to make the assertions work out... so you'd need a bit more to make it work.  I experimented with something like what you're describing in https://github.com/datastax/java-driver/pull/1336/files#r329775123 but it seemed like splitting things out into discrete test cases kept the output readable while still keeping things maintainable.

Reply:
I suppose this isn't _as_ bad now that I look at it:

```java
  @DataProvider
  public static List<List<Object>> createCompressorDP() {

	  Multimap<Class<?>,String> themap = Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
	  themap.putAll(Lz4Compressor.class, ImmutableList.of("lz4", "lZ4", "Lz4", "LZ4"));
	  themap.putAll(SnappyCompressor.class, ImmutableList.of("snappy", "SNAPPY", "sNaPpY", "SNapPy"));
	  themap.putAll(NoopCompressor.class, ImmutableList.of("none", "NONE", "NoNe", "nONe"));
	  
	  return themap.entries().stream()
			  .map(entry -> ImmutableList.<Object>of(Optional.of(entry.getValue()), entry.getKey()))
			  .collect(Collectors.toList());
  }

  @Test
  @UseDataProvider("createCompressorDP")
  public void doCreateCompressorTest(Optional<String> configVal, Class<?> expectedClz) {

    DefaultDriverContext ctx = buildMockedContext(configVal);
    Compressor<ByteBuf> compressor = ctx.getCompressor();
    assertThat(compressor).isNotNull();
    assertThat(compressor).isInstanceOf(expectedClz);
  }
```

@tomekl007 what do you think... do you prefer the all-in-one data provider or discrete providers for discrete test methods?

Reply:
After mulling this over for a bit I think I'll stick with what I have now in the PR.  I think it's easier to read and grok at a glance than the alternatives.  I also like the fact that it highlights the interrelationship of the "noop if undefined" and "noop if explicitly called for" behaviours.  It's also a bit easier to identify individual test cases when running this test in an IDE.  If both params are passed along (i.e. if either of the proposed explicit provider functions are used) the message that gets printed there is kind of ugly.